source/slang/check.cpp

```diff
@@ -140,40 +140,101 @@ namespace Slang
             return result;
         }
 
+        RefPtr<DeclRefType> getExprDeclRefType(Expr * expr)
+        {
+            if (auto typetype = expr->type->As<TypeType>())
+                return typetype->type.As<DeclRefType>();
+            else
+                return expr->type->As<DeclRefType>();
+        }
+
+
         RefPtr<Expr> ConstructDeclRefExpr(
             DeclRef<Decl>   declRef,
             RefPtr<Expr>    baseExpr,
             SourceLoc       loc)
         {
             if (baseExpr)
             {
+                RefPtr<Expr> expr;
+                DeclRef<Decl> *declRefOut;
                 if (baseExpr->type->As<TypeType>())
                 {
-                    auto expr = new StaticMemberExpr();
-                    expr->loc = loc;
-                    expr->BaseExpression = baseExpr;
-                    expr->name = declRef.GetName();
-                    expr->type = GetTypeForDeclRef(declRef);
-                    expr->declRef = declRef;
-                    return expr;
+                    auto sexpr = new StaticMemberExpr();
+                    sexpr->loc = loc;
+                    sexpr->BaseExpression = baseExpr;
+                    sexpr->name = declRef.GetName();
+                    sexpr->declRef = declRef;
+                    declRefOut = &sexpr->declRef;
+                    expr = sexpr;
                 }
                 else
                 {
-                    auto expr = new MemberExpr();
-                    expr->loc = loc;
-                    expr->BaseExpression = baseExpr;
-                    expr->name = declRef.GetName();
-                    expr->type = GetTypeForDeclRef(declRef);
-                    expr->declRef = declRef;
+                    auto sexpr = new MemberExpr();
+                    sexpr->loc = loc;
+                    sexpr->BaseExpression = baseExpr;
+                    sexpr->name = declRef.GetName();
+                    sexpr->declRef = declRef;
+                    declRefOut = &sexpr->declRef;
+                    expr = sexpr;
+                }
+
+                RefPtr<ThisTypeSubstitution> baseThisTypeSubst;
+                if (auto baseDeclRefExpr = baseExpr->As<DeclRefExpr>())
+                {
+                    baseThisTypeSubst = getThisTypeSubst(baseDeclRefExpr->declRef, false);
+                    if (auto baseAssocType = baseDeclRefExpr->declRef.As<AssocTypeDecl>())
```

Comment:
I’m still not 100% happy with these substitutions being introduced here, rather than arising from some lookup-related step (e.g., adding the substitution when we encounter a certain kind of breadcrumb). I’m fine with it for now, but I want to think about possible cleanup down the road.